Implement virtual keyword

Marenz opened the pull request:
refs #5424

Marenz committed:
Parse "virtual" keyword

Marenz committed:
Implement virtual keyword

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `b12074d467808c516988fd18b1280718f0b70024`:
```
Coding style error:
 libsolidity/analysis/ContractLevelChecker.cpp:257: if (function->overrides())

```
Please check that your changes are working as intended.

Marenz commented:
Note: this is currently possible:
```
contract C {
	function() external virtual fp;
	function() external override fp2;
	function() external override virtual fp3;
}
```

Marenz commented:
feedback from the chat:
* dissallow virtual for public variables
* emit warning for virtual in interfaces not only when overriding
* in Token.h, wrong `=>` change

Marenz commented:
Updated:
* hopefully fixed all tests
* added change log
* fixed token.h problem

not fixed: any public variable specific code. I want to fix that behavior in an extra PR

ekpyron commented on test/libsolidity/syntaxTests/viewPureChecker/interface.sol:
```diff
@@ -1,6 +1,8 @@
     function f() override view external {}
 }
+// ----
+// Warning: (18-53): Interface functions are implicitly "virtual"
```
This we don't want to test here, do we?

ekpyron commented on test/libsolidity/syntaxTests/nameAndTypeResolution/425_interface_implement_public_contract.sol:
```diff
@@ -1,7 +1,9 @@
     }
 }
+// ----
+// Warning: (18-48): Interface functions are implicitly "virtual"
```
Same as below - there should be one test for virtual in interfaces and this warning should be avoided anywhere else.

ekpyron commented on test/libsolidity/syntaxTests/nameAndTypeResolution/424_using_interface_complex.sol:
```diff
@@ -9,3 +9,4 @@ abstract contract C is I {
     }
 }
 // ----
+// Warning: (33-63): Interface functions are implicitly "virtual"
```
Remove it.

ekpyron commented on test/libsolidity/syntaxTests/nameAndTypeResolution/359_illegal_override_payable_nonpayable.sol:
```diff
@@ -2,4 +2,5 @@ contract B { function f() public {} }
 contract C is B { function f() payable public {} }
 // ----
 // TypeError: (56-86): Overriding function is missing 'override' specifier.
+// TypeError: (56-86): Trying to override non-virtual function. Did you forget to add "virtual"?
```
Here we should *add* virtual, since the test is about a different reason for the overload to be wrong.

ekpyron commented on test/libsolidity/syntaxTests/nameAndTypeResolution/358_illegal_override_payable.sol:
```diff
@@ -2,4 +2,5 @@ contract B { function f() payable public {} }
 contract C is B { function f() public {} }
 // ----
 // TypeError: (64-86): Overriding function is missing 'override' specifier.
+// TypeError: (64-86): Trying to override non-virtual function. Did you forget to add "virtual"?
```
Also add ``virtual`` here.

ekpyron commented on test/libsolidity/syntaxTests/nameAndTypeResolution/059_illegal_override_visibility.sol:
```diff
@@ -2,4 +2,5 @@ contract B { function f() internal {} }
 contract C is B { function f() public {} }
 // ----
 // TypeError: (58-80): Overriding function is missing 'override' specifier.
+// TypeError: (58-80): Trying to override non-virtual function. Did you forget to add "virtual"?
```
Add ``virtual`` here instead.

ekpyron commented on test/libsolidity/syntaxTests/inheritance/override/remove_view.sol:
```diff
@@ -2,4 +2,5 @@ contract B { function f() public view {} }
 contract C is B { function f() public {} }
 // ----
 // TypeError: (61-83): Overriding function is missing 'override' specifier.
+// TypeError: (61-83): Trying to override non-virtual function. Did you forget to add "virtual"?
```
Also add ``virtual`` here.

ekpyron commented on test/libsolidity/syntaxTests/inheritance/override/override_interface_multiple.sol:
```diff
@@ -0,0 +1,15 @@
+	function test2() external override(B, A) returns (uint256);
+}
+// ----
+// Warning: (168-220): Interface functions are implicitly "virtual"
```
Remove ``virtual``

ekpyron commented on test/libsolidity/syntaxTests/inheritance/override/add_view.sol:
```diff
@@ -2,4 +2,5 @@ contract B { function f() public {} }
 contract C is B { function f() public view {} }
 // ----
 // TypeError: (56-83): Overriding function is missing 'override' specifier.
+// TypeError: (56-83): Trying to override non-virtual function. Did you forget to add "virtual"?
```
Add ``virtual`` here.

ekpyron commented on test/libsolidity/SolidityParser.cpp:
```diff
@@ -118,6 +118,14 @@ while(0)
+BOOST_AUTO_TEST_CASE(reserved_keywords)
+{
+	BOOST_CHECK(!TokenTraits::isReservedKeyword(Token::Identifier));
+	BOOST_CHECK(TokenTraits::isReservedKeyword(Token::Abstract));
```
Curious - this should no longer be reserved, should it? Unrelated to this PR, but we might want to check that!

ekpyron commented on libsolidity/parsing/Parser.h:
```diff
@@ -70,6 +70,7 @@ class Parser: public langutil::ParserBase
 	/// This struct is shared for parsing a function header and a function type.
 	struct FunctionHeaderParserResult
 	{
+		bool isVirtual;
```
Haha, I just wanted to suggest to ditch the ``is`` part, but we, of course, can't just name it ``bool virtual;`` :-D. So fine :-). You might want to member-initialize to ``false``, though.

ekpyron commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -679,6 +689,7 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 
 	bool isIndexed = false;
 	bool isDeclaredConst = false;
+	bool isVirtual = false;
```
Yeah, ok, this is WIP I guess - (and to be changed to not allowing virtual for any variable declarations).

ekpyron replied:
You could still consider to just remove these changes from this PR instead of adding this here only to remove it in a follow-up PR... but as you like!

ekpyron commented on libsolidity/ast/AST.h:
```diff
@@ -619,11 +621,13 @@ class CallableDeclaration: public Declaration, public VariableScope
 	ASTPointer<ParameterList> m_parameters;
 	ASTPointer<OverrideSpecifier> m_overrides;
 	ASTPointer<ParameterList> m_returnParameters;
+	bool m_isVirtual;
```
Member initialize to ``false`` (yeah, it's redundant, but still better to do it and we said we'd want to always do it for PODs).

ekpyron commented on libsolidity/ast/AST.h:
```diff
@@ -814,6 +822,7 @@ class VariableDeclaration: public Declaration
 	bool m_isStateVariable; ///< Whether or not this is a contract state variable
 	bool m_isIndexed; ///< Whether this is an indexed variable (used by events).
 	bool m_isConstant; ///< Whether the variable is a compile-time constant.
+	bool m_isVirtual; ///< Whether the variable is virtual and can be overridden
```
Haha, I *would* suggest to (redundantly) member-initialize this to ``false`` as well, but looking at the lines above this would be somewhat superficial :-D. Can you just member-initialize all of these ``bool``s to ``false`` anyways and also ``m_location`` below to ``Location::Unspecified``? We have to start doing it consistently some time, so why not now...

ekpyron replied:
I haven't done this consistently myself in the recent past, though, I must admit, since I just read the not-member-initialized ``FunctionDefinition::kind`` above :-). But still.

ekpyron commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -279,6 +284,12 @@ bool ContractLevelChecker::checkFunctionOverride(FunctionDefinition const& _func
 		success = false;
 	}
 
+	if (!_super.isVirtual() && !_super.annotation().contract->isInterface())
```
We could even think about adding a ``isVirtualOrInterfaceFunction`` that checks both - that might make it less likely to miss the fact that interface functions are implicitly virtual despite ``isVirtual`` returning ``false``? But yeah: as you like, I'd be fine without that.

Marenz replied:
Would you then get rid of `isVirtual`?

chriseth replied:
`hasVirtualSemantics` - `isMarkedVirtual`

chriseth replied:
Or better in line with coding stile: `virtualSemantics` - `markedVirtual`

ekpyron replied:
Or maybe this will be the new ``canBeOverridden`` :-)? But yeah, I think it'd be best to split into two, but I don't mind the exact naming - e.g. ``virtualSemantics`` and ``markedVirtual`` would work for me.

ekpyron replied:
Or ``overridable()`` for that matter - but in the end that's the semantics of being ``virtual``, isn't it?

ekpyron commented on liblangutil/Token.h:
```diff
@@ -313,7 +313,7 @@ namespace TokenTraits
 	constexpr bool isEtherSubdenomination(Token op) { return op == Token::SubWei || op == Token::SubSzabo || op == Token::SubFinney || op == Token::SubEther; }
 	constexpr bool isTimeSubdenomination(Token op) { return op == Token::SubSecond || op == Token::SubMinute || op == Token::SubHour || op == Token::SubDay || op == Token::SubWeek || op == Token::SubYear; }
-	constexpr bool isReservedKeyword(Token op) { return (Token::Abstract <= op && op <= Token::Virtual); }
+	constexpr bool isReservedKeyword(Token op) { return (Token::Abstract <= op && op <= Token::Unchecked); }
```
If you want you can move ``Abstract`` out of the reserved keywords as well here - otherwise we should do it separately - apparently that was missed in the abstract PR. Or is this one just not rebased on top of that one?

Marenz replied:
I believe it was forgotten. I'll move it.

ekpyron reviewed (changes requested):
This looks good - only some minor suggestions apart from the tests - for the tests we should take care that they - as far as possible - check what they are supposed to and *only* check what they are supposed to.

chriseth commented on Changelog.md:
```diff
@@ -21,6 +21,8 @@ Breaking changes:
 
 
  * Language Feature: When overriding a function or modifier, the new keyword ``override`` must be used. When overriding a function or modifier defined in multiple parallel bases, all bases must be listed in parentheses after the keyword like so: ``override(Base1, Base2)``
+ * Language Feature: Functions can only be overridden when they are marked with
```
I think this still fits the first line

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -245,19 +245,24 @@ void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _cont
-				"Function has override specified but does not override anything."
-			);
+		if (!funcSet.count(function))
+		{
```
`{` not needed.

chriseth replied:
Actually not even nested if needed.

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -279,6 +284,12 @@ bool ContractLevelChecker::checkFunctionOverride(FunctionDefinition const& _func
 
+	if (!_super.isVirtual() && !_super.annotation().contract->isInterface())
+	{
+		overrideError(_function, _super, "Trying to override non-virtual function. Did you forget to add \"virtual\"?");
```
Is the error shown on `_super` or on `_function`? Should be shown on `_super`.

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -781,8 +792,7 @@ void ContractLevelChecker::checkOverrideList(FunctionMultiSet const& _funcSet, F
 	{
 		// Validate the override
-		if (!checkFunctionOverride(_function, **begin))
-			break;
```
Does this generate duplicate errors now?

Marenz replied:
well, it can print similar errors for overrides, but the secondary error location would then show to a different overloaded function.

Example:

```
abstract contract C
{
	function foo() internal;
}
abstract contract A
{
	function foo() internal;
}
contract D is C, A {
	function foo() internal override(A, C) {}
}

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:10:2:
  |
10 |    function foo() internal override(A, C) {}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overridden function is here:
 --> test.sol:3:2:
  |
3 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:10:2:
  |
10 |    function foo() internal override(A, C) {}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overridden function is here:
 --> test.sol:7:2:
  |
7 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^


```

chriseth replied:
Maybe it would be better to swap primary and secondary locations?

chriseth replied:
Ok, maybe the order is correct, but we could move the `Did you forget to add "virtual"` to the secondary location? As it is now, it could suggest that you should add `virtual` to the overriding function.

Marenz replied:
That has already been done here: https://github.com/ethereum/solidity/pull/7647/files#diff-b1ba2bbc8a967e12c1721caf45edb970R283

The example is from an earlier time

Marenz replied:
For clarification, this is how the output with the current code looks:

```
Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:4:2:
  |
4 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overriding function is here:
  --> test.sol:11:2:
   |
11 |    function foo() internal override(A, C) {}
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:8:2:
  |
8 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overriding function is here:
  --> test.sol:11:2:
   |
11 |    function foo() internal override(A, C) {}
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -804,13 +824,32 @@ ASTPointer<ModifierDefinition> Parser::parseModifierDefinition()
 	ASTPointer<OverrideSpecifier> overrides;
+	bool isVirtual = false;
+
+	while(true)
```
```suggestion
	while (true)
```

ekpyron replied:
This should be added to ``check_style.sh`` I guess...

chriseth commented:
Please two tests where C inherits from contracts A, B, and both A and B define a function that is virtual in A and non-virtual in B and

1. the function is defined in C
2. the function is not defined in C

Marenz committed:
Make ``abstract`` keyword instead of reserved

Marenz commented:
uuuuu[..]uuupdated

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `e834982b117ef34cd10e46584bb9487f65755259`:
```
Coding style error:
 libsolidity/analysis/StaticAnalyzer.cpp:232: while(expr)

```
Please check that your changes are working as intended.

Marenz commented:
moved rule addition to own PR: https://github.com/ethereum/solidity/pull/7683

Marenz commented:
rebased & pushed to get the CI fixes

Marenz commented:
Hmm seems we're still missing one test fix

ekpyron commented:
@Marenz could you rebase again? That should now fix the tests, shouldn't it?

Marenz commented:
Updated changelog
updated breaking changes descr.
moved warning for virtual in interfaces to type checker

ekpyron commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -466,6 +466,7 @@ Parser::FunctionHeaderParserResult Parser::parseFunctionHeader(bool _isStateVari
 	RecursionGuard recursionGuard(*this);
 	FunctionHeaderParserResult result;
 
+	result.isVirtual = false;
```
This line isn't really necessary anymore now that we have it member-initialized... Do you still want to keep it? For this line I'd be fine with keeping it...
Actually the line below is worse and what caught my eye here :-), because it suggests ``result.overrides`` is a plain pointer, which it isn't... it so happens that ``shared_ptr`` has a constructor for ``nullptr_t``, so it works, but if anything, I'd rather use ``result.overrides = {};`` or ``result.overrides.reset();``... or actually remove the line...

So yeah, I'd suggest to remove this line and the one below altogether... what do you think?

Marenz replied:
I am fine with removing both lines.

Marenz replied:
updated

ekpyron reviewed (commented):
This looks good to me now! One minor comment and I haven't yet checked, whether we have sufficient test coverage now. Also I haven't looked at the docs in detail yet.

Marenz commented:
All change requests have been made, this is till waiting for the final review
